processor/k8sprocessor/kube/owner.go

```diff
@@ -0,0 +1,324 @@
+}
+
+func (op *OwnerCache) cacheMetadataObject(kind string, obj *meta_v1.ObjectMeta) {
+	oo := ObjectOwner{
+		UID:       obj.UID,
+		namespace: obj.Namespace,
+		ownerUIDs: []types.UID{},
+		kind:      kind,
+		name:      obj.Name,
+	}
+	for _, or := range obj.OwnerReferences {
+		oo.ownerUIDs = append(oo.ownerUIDs, or.UID)
+	}
+
+	// First set the cache
+	op.objectOwnersCache.Add(string(obj.UID), &oo, gocache.DefaultExpiration)
+}
+
+type objectOrList struct {
+	obj  *meta_v1.ObjectMeta
+	list []*meta_v1.ObjectMeta
+}
+
+// objectApiCall queries for ObjectMeta; if name is nil, then all objects of given type from
+// namespace are retrieved; otherwise only the one with specified name. Errors are quietly ignored
+func (op *OwnerCache) objectApiCall(namespace string, kind string, name *string) objectOrList {
+	result := objectOrList{}
+	result.list = make([]*meta_v1.ObjectMeta, 0)
+
+	startTime := time.Now()
+	callNotMade := false
+
+	// This *terribly* misses that: https://blog.golang.org/why-generics
+	switch kind {
+	case "DaemonSet":
+		daemonSetAPI := op.clientset.ExtensionsV1beta1().DaemonSets(namespace)
+
+		if name != nil {
+			ds, _ := daemonSetAPI.Get(*name, meta_v1.GetOptions{})
+			if ds != nil {
+				result.obj = &ds.ObjectMeta
+			}
+		} else {
+			dss, _ := daemonSetAPI.List(meta_v1.ListOptions{})
+			if dss != nil {
+				result.list = make([]*meta_v1.ObjectMeta, len(dss.Items))
+				for i, it := range dss.Items {
+					result.list[i] = &it.ObjectMeta
+				}
+			}
+		}
+	case "Deployment":
+		deploymentsAPI := op.clientset.ExtensionsV1beta1().Deployments(namespace)
+
+		if name != nil {
+			ds, _ := deploymentsAPI.Get(*name, meta_v1.GetOptions{})
+			if ds != nil {
+				result.obj = &ds.ObjectMeta
+			}
+		} else {
+			dss, _ := deploymentsAPI.List(meta_v1.ListOptions{})
+			if dss != nil {
+				result.list = make([]*meta_v1.ObjectMeta, len(dss.Items))
+				for i, it := range dss.Items {
+					result.list[i] = &it.ObjectMeta
+				}
+			}
+		}
+	case "Pod":
+		podsAPI := op.clientset.CoreV1().Pods(namespace)
+
+		if name != nil {
+			ds, _ := podsAPI.Get(*name, meta_v1.GetOptions{})
+			if ds != nil {
+				result.obj = &ds.ObjectMeta
+			}
+		} else {
+			dss, _ := podsAPI.List(meta_v1.ListOptions{})
+			if dss != nil {
+				result.list = make([]*meta_v1.ObjectMeta, len(dss.Items))
+				for i, it := range dss.Items {
+					result.list[i] = &it.ObjectMeta
+				}
+			}
+		}
+	case "ReplicaSet":
+		replicaSetsAPI := op.clientset.ExtensionsV1beta1().ReplicaSets(namespace)
+
+		if name != nil {
+			ds, _ := replicaSetsAPI.Get(*name, meta_v1.GetOptions{})
+			if ds != nil {
+				result.obj = &ds.ObjectMeta
+			}
+		} else {
+			dss, _ := replicaSetsAPI.List(meta_v1.ListOptions{})
+			if dss != nil {
+				result.list = make([]*meta_v1.ObjectMeta, len(dss.Items))
+				for i, it := range dss.Items {
+					result.list[i] = &it.ObjectMeta
+				}
+			}
+		}
+	case "Service":
+		servicesAPI := op.clientset.CoreV1().Services(namespace)
```

Comment:
That's a very fair question, I think this is a viable scenario (though probably infrequent). I need to research it bit more. The endpoint address would change in such case, but it is not being used for assigning the metadata anyways...

Reply:
I looked into it and made a quick test and it seems there's really no easy way to attribute the proper Service name in case several of those are connected. In some cases, it could be extracted from the URL (e.g. for http) but I believe adding such capability would be somewhat complex and not worth the cost. Hence - maybe we should leave it as a current limitation?

Reply:
I refactored this significantly and change the way services are retrieved. If there are several of them pointing to the pod, they are now comma-separated